isis/src/base/apps/csminit/csminit.cpp

```diff
@@ -46,101 +48,188 @@ namespace Isis {
+          QString currentPluginName = QString::fromStdString(plugin->getPluginName());
+          for (size_t modelIndex = 0; modelIndex < plugin->getNumModels(); modelIndex++) {
+            QString modelName = QString::fromStdString(plugin->getModelName(modelIndex));
+            message += "Plugin [" + currentPluginName + "], Model [" + modelName + "]\n";
+          }
+        }
+        throw IException(IException::User, message, _FILEINFO_);
       }
-      throw IException(IException::User, message, _FILEINFO_);
-    }
 
-    if (possibleModels.empty()) {
-      QString message = "No loaded model could be created from the ISD [" + isdFilePath + "]."
-                        "Loaded plugin & model names:\n";
-      for (const csm::Plugin * plugin : csm::Plugin::getList()) {
-        QString pluginName = QString::fromStdString(plugin->getPluginName());
-        for (size_t modelIndex = 0; modelIndex < plugin->getNumModels(); modelIndex++) {
-          QString modelName = QString::fromStdString(plugin->getModelName(modelIndex));
-          message += "Plugin [" + pluginName + "], Model [" + modelName + "]\n";
-        }
+      // If we are here, then we have exactly 1 model
+      QStringList modelSpec = possibleModels.front();
+
+      if (modelSpec.size() != 3) {
+        QString message = "Model specification [" + modelSpec.join(" ") + "] has [" + modelSpec.size() + "] elements "
+          "when it should have 3 elements.";
+        throw IException(IException::Programmer, message, _FILEINFO_);
+      }
+
+      pluginName = modelSpec[0];
+      modelName = modelSpec[1];
+      QString isdFormat = modelSpec[2];
+
+      const csm::Plugin *plugin = csm::Plugin::findPlugin(pluginName.toStdString());
+      if (plugin == NULL) {
+        QString message = "Cannot find requested Plugin: [" + pluginName + "].";
+        throw IException(IException::User, message, _FILEINFO_);
       }
+
+      csm::Isd fileIsd(isdFilePath.toStdString());
+      csm::Nitf21Isd nitf21Isd(isdFilePath.toStdString());
+      if (isdFormat == QString::fromStdString(fileIsd.format())) {
+        model = plugin->constructModelFromISD(fileIsd, modelName.toStdString());
+      }
+      else if (isdFormat == QString::fromStdString(nitf21Isd.format())) {
+        model = plugin->constructModelFromISD(nitf21Isd, modelName.toStdString());
+      }
+      else {
+        QString message = "Invalid ISD format specifications [" + isdFormat + "].";
+        throw IException(IException::Programmer, message, _FILEINFO_);
+      }
+    } // end of ISD if statement
+
+    else if (ui.WasEntered("STATE")) {
+      FileName stateFilePath = ui.GetFileName("STATE");
+
+      std::ifstream file(stateFilePath.expanded().toStdString());
+      std::stringstream buffer;
+      buffer << file.rdbuf();
+      QString stateString = QString::fromStdString(buffer.str());
+
+      if (!ui.WasEntered("PLUGINNAME") && !ui.WasEntered("MODELNAME")) {
```

Comment:
The else on this if / else block handles the neither case. I'll move the else up to here and do an explicit check for neither so that the two errors are next to each other.